src/typecheck.c

```diff
@@ -631,17 +632,40 @@ static const char *nth(int n)
 }
 
 // returns NULL if the function doesn't return anything, otherwise non-owned pointer to non-owned type
-static const Type *typecheck_function_call(TypeContext *ctx, const AstCall *call, Location location)
+static const Type *typecheck_function_or_method_call(TypeContext *ctx, const AstCall *call, const Type *self_type, Location location)
 {
-    const Signature *sig = find_function(ctx, call->calledname);
-    if (!sig)
-        fail_with_error(location, "function '%s' not found", call->calledname);
+    const Signature *sig;
+    char function_or_method[20];
+
+    if (self_type) {
+        strcpy(function_or_method, "method");
+
+        if (self_type->kind != TYPE_STRUCT) {
+            fail_with_error(location,
+                "type %s does not have a method named '%s', and in fact, it doesn't have any methods because it is not a struct",
+                self_type->name, call->calledname);
+        }
+
+        char name_to_search[sizeof self_type->name + sizeof "." + sizeof call->calledname];
+        sprintf(name_to_search, "%s.%s", self_type->name, call->calledname);
+        sig = find_function(ctx, name_to_search);
+        if (!sig)
+            fail_with_error(location, "struct %s does not have a method named '%s'", self_type->name, call->calledname);
+    } else {
+        strcpy(function_or_method, "function");
+        sig = find_function(ctx, call->calledname);
+        if (!sig)
+            fail_with_error(location, "function '%s' not found", call->calledname);
+    }
+
     char *sigstr = signature_to_string(sig, false);
 
-    if (call->nargs < sig->nargs || (call->nargs > sig->nargs && !sig->takes_varargs)) {
+    int n = call->nargs + !!self_type;
+    if (n < sig->nargs || (n > sig->nargs && !sig->takes_varargs)) {
         fail_with_error(
             location,
-            "function %s takes %d argument%s, but it was called with %d argument%s",
+            "%s %s takes %d argument%s, but it was called with %d argument%s",
```

Comment:
Should add a test for this on methods. And probably for everything else that mentions a "function", as that should automagically say "method" instead